src/ws/test-channelresponse.c

```diff
@@ -525,19 +527,24 @@ test_resource_failure (TestResourceCase *tc,
   cockpit_expect_possible_log ("cockpit-protocol", G_LOG_LEVEL_WARNING, "*: bridge program failed:*");
   cockpit_expect_possible_log ("cockpit-ws", G_LOG_LEVEL_MESSAGE, "*: external channel failed: *");
 
-  /* Now kill the bridge */
+  /* Make a pidfd for the bridge */
   g_assert (cockpit_pipe_get_pid (tc->pipe, &pid));
   g_assert_cmpint (pid, >, 0);
+  int pid_fd = syscall(__NR_pidfd_open, pid, 0);
+
+  /* Now kill the bridge */
   g_assert_cmpint (kill (pid, SIGTERM), ==, 0);
-  /* Wait until it's gone */
-#if GLIB_CHECK_VERSION(2,73,2) && !defined(__MIPSEL)
-  /* https://gitlab.gnome.org/GNOME/glib/-/commit/f615eef4bafaa2f dropped global GChildWatch, we need to wait ourselves */
-  g_assert_cmpint (waitpid (pid, NULL, 0), ==, pid);
-#else
-  /* on older glib versions, waitpid interferes with GChildWatch */
-  while (kill (pid, 0) >= 0)
-    g_usleep (1000);
-#endif
+
+  /* The SIGTERM gets delivered to the bridge via a glib unix signal
+   * handler, and it is theoretically possible that the request that we
+   * send below could get delivered before the SIGTERM.  For that
+   * reason, we need to make sure that the process actually properly
+   * exited before sending the request.
+   */
+  struct pollfd pid_pfd = { .fd = pid_fd, .events = POLLIN };
+  while (poll (&pid_pfd, 1, -1) != 1)
+    ;
```

Comment:
Could this do the `g_usleep (1000)` again to not spin in a tight loop (which would only make the process shut down more slowly)?

Reply:
This isn't spinning a tight loop.  `-1` as the last parameter to `poll()` means "sleep indefinitely".

Reply:
Nevermind -- poll timeout does that.

Reply:
The loop is more about dealing with the (in this case extremely likely) `EINTR` case when the `SIGCHLD` gets delivered to us and restarting the call.  I could have made that more explicit, but this is also just a test...